src/sdf/include/sdf.h

```diff
@@ -0,0 +1,474 @@
+      : step_{step}, table_{alloc}, used_{used} {}
+
+  __host__ __device__ int Size() const { return table_.size(); }
+
+  __host__ __device__ bool Insert(const GridVert& vert) {
+    uint32_t idx = vert.key & (Size() - 1);
+    while (1) {
+      Uint64& key = table_[idx].key;
+      const Uint64 found = AtomicCAS(key, kOpen, vert.key);
+      if (found == kOpen) {
+        if (AtomicAdd(used_[0], 0x1u) * 2 > Size()) {
+          return true;
+        }
+        table_[idx] = vert;
+        return false;
+      }
+      if (found == vert.key) return false;
+      idx = (idx + step_) & (Size() - 1);
+    }
+  }
+
+  __host__ __device__ GridVert operator[](Uint64 key) const {
+    uint32_t idx = key & (Size() - 1);
+    while (1) {
+      const GridVert found = table_[idx];
+      if (found.key == key || found.key == kOpen) return found;
+      idx = (idx + step_) & (Size() - 1);
+    }
+  }
+
+  __host__ __device__ GridVert At(int idx) const { return table_[idx]; }
+
+ private:
+  const uint32_t step_;
+  VecD<GridVert> table_;
+  VecD<uint32_t> used_;
+};
+
+class HashTable {
+ public:
+  HashTable(uint32_t size, uint32_t step = 127)
+      : alloc_{1 << (int)ceil(log2(size)), {}}, table_{alloc_, used_, step} {}
+
+  HashTableD D() { return table_; }
+
+  int Entries() const { return used_[0]; }
+
+  int Size() const { return table_.Size(); }
+
+  float FilledFraction() const {
+    return static_cast<float>(used_[0]) / table_.Size();
+  }
+
+ private:
+  VecDH<GridVert> alloc_;
+  VecDH<uint32_t> used_ = VecDH<uint32_t>(1, 0);
+  HashTableD table_;
+};
+
+template <typename Func>
+struct ComputeVerts {
+  glm::vec3* vertPos;
+  int* vertIndex;
+  HashTableD gridVerts;
+  const Func sdf;
+  const glm::vec3 origin;
+  const glm::ivec3 gridSize;
+  const glm::vec3 spacing;
+  const float level;
+
+  inline __host__ __device__ glm::vec3 Position(glm::ivec4 gridIndex) const {
+    return origin + spacing * (-0.5f + glm::vec3(gridIndex) +
+                               (gridIndex.w == 1 ? 0.5f : 0.0f) * glm::vec3(1));
+  }
+
+  inline __host__ __device__ float BoundedSDF(glm::ivec4 gridIndex) const {
+    const float d = sdf(Position(gridIndex)) - level;
+
+    const glm::ivec3 xyz(gridIndex);
+    if (glm::any(glm::equal(xyz, glm::ivec3(0))) ||
+        glm::any(glm::greaterThanEqual(xyz, gridSize)) ||
+        (gridIndex.w == 1 &&
+         glm::any(glm::greaterThanEqual(xyz, gridSize - 1))))
+      return glm::min(d, 0.0f);
+
+    return d;
+  }
+
+  inline __host__ __device__ void operator()(Uint64 mortonCode) {
+    const glm::ivec4 gridIndex = DecodeMorton(mortonCode);
+
+    if (glm::any(glm::greaterThan(glm::ivec3(gridIndex), gridSize))) return;
+
+    const glm::vec3 position = Position(gridIndex);
+
+    GridVert gridVert;
+    gridVert.key = mortonCode;
+    gridVert.distance = BoundedSDF(gridIndex);
+
+    bool keep = false;
+    float minDist2 = 0.25 * 0.25;
```

Comment:
What is the usage of `minDist2`?

Reply:
Ah yes, that's dead code now. I had a cool idea, but realized it would create edge cases that could break manifoldness.